File: Main/src/main/java/com/redelf/commons/persistance/Storage.java
package com.redelf.commons.persistance;

import com.redelf.commons.lifecycle.InitializationWithContext;
import com.redelf.commons.lifecycle.ShutdownSynchronized;
import com.redelf.commons.lifecycle.TerminationSynchronized;

public interface Storage<T> extends

        ShutdownSynchronized,
        TerminationSynchronized,
        InitializationWithContext

{

  boolean put(String key, T value);

  T get(String key);

  boolean delete(String key);

  boolean deleteAll();

  long count();

  boolean contains(String key);
}

File: Main/src/main/java/com/redelf/commons/persistance/base/Storage.kt
package com.redelf.commons.persistance.base

import com.redelf.commons.lifecycle.initialization.InitializationWithContext
import com.redelf.commons.lifecycle.shutdown.ShutdownSynchronized
import com.redelf.commons.lifecycle.termination.TerminationSynchronized
import com.redelf.commons.obtain.OnObtain

/*
    KNOWN LIMITATION: This interface extends both ShutdownSynchronized and TerminationSynchronized,
    which have overlapping responsibilities. A future refactoring should consolidate these into
    a single lifecycle interface across the codebase.
*/
interface Storage<T> : ShutdownSynchronized, TerminationSynchronized, InitializationWithContext {

    fun put(key: String?, value: T): Boolean

    fun get(key: String?, callback: OnObtain<T?>)

    fun delete(key: String?): Boolean

    fun deleteAll(): Boolean

    fun count(): Long

    fun contains(key: String?, callback: OnObtain<Boolean?>)
}
